src/substrait/builders/extended_expression.py

```diff
@@ -0,0 +1,374 @@
+                                struct_field=stalg.Expression.ReferenceSegment.StructField(
+                                    field=field_index
+                                )
+                            ),
+                        )
+                    ),
+                    output_names=list(base_schema.names)[names_start:names_end]
+                    if not alias
+                    else [alias],
+                )
+            ],
+            base_schema=base_schema,
+        )
+
+    return resolve
+
+def scalar_function(
+    uri: str, function: str, *expressions: UnboundExtendedExpression, alias: str = None
+):
+    """Builds a resolver for ExtendedExpression containing a ScalarFunction expression"""
+
+    def resolve(
+        base_schema: stp.NamedStruct, registry: ExtensionRegistry
+    ) -> stee.ExtendedExpression:
+        bound_expressions: Iterable[stee.ExtendedExpression] = [
+            e(base_schema, registry) for e in expressions
+        ]
+
+        expression_schemas = [
+            infer_extended_expression_schema(b) for b in bound_expressions
+        ]
+
+        signature = [typ for es in expression_schemas for typ in es.types]
+
+        func = registry.lookup_function(uri, function, signature)
+
+        if not func:
+            raise Exception(f"Unknown function {function} for {signature}")
+
+        func_extension_uris = [
+            ste.SimpleExtensionURI(
+                extension_uri_anchor=registry.lookup_uri(uri), uri=uri
+            )
+        ]
+
+        func_extensions = [
+            ste.SimpleExtensionDeclaration(
+                extension_function=ste.SimpleExtensionDeclaration.ExtensionFunction(
+                    extension_uri_reference=registry.lookup_uri(uri),
+                    function_anchor=func[0].anchor,
+                    name=function,
+                )
+            )
+        ]
+
+        extension_uris = merge_extension_uris(
+            func_extension_uris, *[b.extension_uris for b in bound_expressions]
+        )
+
+        extensions = merge_extension_declarations(
+            func_extensions, *[b.extensions for b in bound_expressions]
+        )
+
+        return stee.ExtendedExpression(
+            referred_expr=[
+                stee.ExpressionReference(
+                    expression=stalg.Expression(
+                        scalar_function=stalg.Expression.ScalarFunction(
+                            function_reference=func[0].anchor,
+                            arguments=[
+                                stalg.FunctionArgument(
+                                    value=e.referred_expr[0].expression
+                                )
+                                for e in bound_expressions
+                            ],
+                            output_type=func[1],
+                        )
+                    ),
+                    output_names=[alias if alias else "scalar_function"],
```

Comment:
Should this be a unique name?  Perhaps keep track of generated names and prepend a count for each reuse (and elsewhere).

Reply:
I switched to a rudimentary way of generating names from argument names. (ibis does something like this) Pretty sure it's gonna fail if the output type of an expression contains a struct, but it's a start. I'm not sure if it's worth it to infer an output type just to get the expected cardinality of the `output_names` array.